sycl/doc/design/CommandGraph.md

```diff
@@ -1,66 +1,164 @@
+| `urCommandBufferAppendKernelLaunchExp`  | Append a kernel execution command to command-buffer. |
+| `urCommandBufferAppendMemcpyUSMExp`     | Append a USM memcpy command to the command-buffer. |
+| `urCommandBufferAppendMembufferCopyExp` | Append a mem buffer copy command to the command-buffer. |
+| `urCommandBufferEnqueueExp`             | Submit command-buffer to a command-queue for execution. |
+
+See the [UR EXP-COMMAND-BUFFER](https://oneapi-src.github.io/unified-runtime/core/EXP-COMMAND-BUFFER.html)
+specification for more details.
 
 ## Design
 
 ![Basic architecture diagram.](images/SYCL-Graph-Architecture.svg)
 
 There are two sets of user facing interfaces that can be used to create a
-command graph object: 
-Explicit and Record & Replay API. Within the runtime they share a common
-infrastructure.
+command-graph object: Explicit and Record & Replay API. Within the runtime they
+share a common infrastructure.
+
+## Nodes & Edges
+
+Nodes are represented by an internal `detail::node_impl` class which
+stores a `sycl::detail::CG` object for the command-group that the node
+represents. When a user adds a node to a graph using the explicit
+`command_graph<modifiable>::add()` API passing a CGF, in the implementation a
+`sycl::handler` object is created which finalizes the CGF, and after
+finalization the CG object from the handler is moved to the node.
+
+For creating a node in the graph using queue recording mode. When the
+`sycl::handler` from a queue submissions is finalized, if the queue the
+handler was created from is in the recording mode, then the command-group from
+finalization is added to the graph associated with the queue as a new node.
+
+Edges are stored in each node as lists of predecessor and successor nodes.
+
+## Sorting
+
+The current way graph nodes are linearized into execution order is using a
+reversed depth-first algorithm. Alternative algorithms, such as breadth-first,
+are possible and may give better performance on certain workloads/hardware. We
+are looking into giving the user control of this implementation detail.
+
+## Scheduler Integration
 
-## Scheduler integration
+When there are no requirements from accessors in a command-graph submission,
+the scheduler is bypassed and the underlying UR command-buffer is directly
+enqueued to a UR queue. If there are accessor requirements, the UR
+command-buffer for the executable graph needs to be enqueued by the scheduler.
 
-When there are no requirements for accessors in a command graph the scheduler
-is bypassed and it is directly enqueued to a command buffer. If 
-there are requirements, commands need to be enqueued by the scheduler. 
+When individual graph nodes have requirements from SYCL accessors, the
+underlying `sycl::detail::CG` object stored in the node is copied and passed to
+the scheduler for adding to the UR command-buffer, otherwise the node can
+be appended directly as a command in the UR command-buffer.
```

Comment:
Might be worth noting here that this is in keeping with the existing behaviour of the handler/normal queue submissions.